### PR TITLE
stalemate implementation

### PR DESCRIPTION
**Pull request by tsaviojojo**
implemented stalemate

### CODE REVIEW

File: src/main/java/jchess/ruleengine/DefaultRuleProcessor.java [hunk 1]
```diff
@@ -432,4 +432,97 @@ public IPlayerAgent tryCheckIfPlayerEndengered(IBoardAgent oBoard, IPlayerAgent
...
+	 * The rules of chess provide that when stalemate occurs, the game ends as a draw.
+	 */
+	
+	public Boolean checkStalemate(IBoardAgent oBoard, IPlayerAgent oPlayer) {
+		Boolean bStalemateCheckIfPlayerEndengered = false;
+		Boolean bStalemateCheckIfPlayNotPossible = false;
+		int nCountNoOfPieces =0;
+		int nCountPiecesTrapped=0;
+		int nCounterOfKingMoves=0;
+		int[] iaArrayOfPossibleChecks = new int[]{0,0,0,0,0,0,0,0,0};
```
Comment (SajadKarim) [inline src/main/java/jchess/ruleengine/DefaultRuleProcessor.java]: Could you please tell me what this array is taken for? It would be helpful if you write your logic that how you achieved the goal.

File: src/main/java/jchess/ruleengine/DefaultRuleProcessor.java [hunk 2]
```diff
@@ -131,28 +131,28 @@ public void tryFindPossibleCandidateMovePositions(IPieceAgent oPieceToMove, IPos
...
 	 * Fig 3: 2-Players chess board.
 	 * 
-	 * Consider Fig 3 (2 Player chess board) and cell “C3” is the active position, and the game engine has to find all the possible moves using
+	 * Consider Fig 3 (2 Player chess board) and cell "C3" is the active position, and the game engine has to find all the possible moves using
 	 * Blinker manoeuvre strategy.
 	 * 
-	 * Cell “C3” is surrounded by 8 cells; the cells are B3, B2, C2, D2, D3, D4, C4, B4 and they are in direction West, SouthWest, South, 
-	 * SouthEast, East, NorthEast, North, NorthWest respectively. All this information is stored in elements “Directions” and “Connections”
-	 * under the element “Position” (Refer to positions section of XML for details).
+	 * Cell â€œC3â€� is surrounded by 8 cells; the cells are B3, B2, C2, D2, D3, D4, C4, B4 and they are in direction West, SouthWest, South,
```
Comment (SajadKarim) [inline src/main/java/jchess/ruleengine/DefaultRuleProcessor.java]: Please revert these changes.

File: src/main/java/jchess/ruleengine/DefaultRuleProcessor.java [hunk 3]
```diff
@@ -432,4 +432,97 @@ public IPlayerAgent tryCheckIfPlayerEndengered(IBoardAgent oBoard, IPlayerAgent
...
+					if( oCurrentPlayer == null)
+						bStalemateCheckIfPlayerEndengered = true;
+					
+					Map<String, IMoveCandidate> mpCandidateMovePositions = new HashMap<String, IMoveCandidate>();
+					tryEvaluateAllRules(oBoard, oRandomPiece, mpCandidateMovePositions);					
+					
+					//Iterating the different moves that the king can make
+					for(Map.Entry<String, IMoveCandidate> oCandidateMovePostion : mpCandidateMovePositions.entrySet()) {
+						
+						int iKingPieceRank = oCandidateMovePostion.getValue().getCandidatePosition().getRank();
```
Comment (SajadKarim) [inline src/main/java/jchess/ruleengine/DefaultRuleProcessor.java]: When there is int, prefix should be n instead of i. E.g. nKingPieceRank

File: src/main/java/jchess/ruleengine/DefaultRuleProcessor.java [hunk 4]
```diff
@@ -432,4 +432,97 @@ public IPlayerAgent tryCheckIfPlayerEndengered(IBoardAgent oBoard, IPlayerAgent
...
+		//fetching all pieces from the board
+		for(Map.Entry<String,IPositionAgent> oPositionPiece : oBoard.getAllPositionAgents().entrySet()) 
+		{
+			IPieceAgent oRandomPiece = oPositionPiece.getValue().getPiece();
+			//check to filter out only the pieces of the current player
+			if(oRandomPiece != null && oRandomPiece.getPlayer()== oPlayer) 
+			{
+				nCountNoOfPieces+=1;
+				//checking if the piece iterated is a king
+				if(oRandomPiece.getPieceData().getName().startsWith("King"))
```
Comment (SajadKarim) [inline src/main/java/jchess/ruleengine/DefaultRuleProcessor.java]: 1. You can directly get name from oRandomPiece.getName().
2. Danny has added a function in it. Try oPlayer.getKingPiece()

File: src/main/java/jchess/ruleengine/DefaultRuleProcessor.java (hunk 1 above)
Comment (tsaviojojo) [reply src/main/java/jchess/ruleengine/DefaultRuleProcessor.java]: this array is taken for the movements that the king can make (max movement is 8) and we find out the possible movement that can lead to checkmate. If all the movements of the king leads to checkmate and the current position is not a check and no other pieces of the current player can be moved then its a stalemate resulting in a draw

File: src/main/java/jchess/ruleengine/DefaultRuleProcessor.java (hunk 4 above)
Comment (tsaviojojo) [reply src/main/java/jchess/ruleengine/DefaultRuleProcessor.java]: Changed it to oRandomPiece.getName().
